File: source/SharpFlame/PathfinderConnection.cs
using System.Diagnostics;

namespace SharpFlame
{
    public class PathfinderConnection
    {
        public bool Destroyed;

        public int Layer_ConnectionNum = -1;

        public PathfinderNode NodeA;

        public PathfinderNode GetNodeA
        {
            get { return NodeA; }
        }

        public int NodeA_ConnectionNum = -1;

        public int GetNodeA_ConnectionNum
        {
            get { return NodeA_ConnectionNum; }
        }

        public PathfinderNode NodeB;

        public PathfinderNode GetNodeB
        {
            get { return NodeB; }
        }

        public int NodeB_ConnectionNum = -1;

        public int GetNodeB_ConnectionNum
        {
            get { return NodeB_ConnectionNum; }
        }

        private PathfinderConnection DependantConnection; //the one above this that partially relies on this to exist

        private int LinkCount;

        public float Value = 1.0F;

        public float GetValue
        {
            get { return this.Value; }
        }

        public int CalcValueNum = -1;

        public PathfinderConnection(PathfinderNode NewNodeA, PathfinderNode NewNodeB, float NewValue)
        {
            if ( NewNodeA.Layer.Network_LayerNum > 0 | NewNodeB.Layer.Network_LayerNum > 0 | NewValue <= 0.0F )
            {
                Debugger.Break();
                return;
            }

            Value = NewValue;

            LinkCount = 1;

            NodeA = NewNodeA;
            NodeB = NewNodeB;
            NodeA.Connection_Add(this, ref NodeA_ConnectionNum);
            NodeB.Connection_Add(this, ref NodeB_ConnectionNum);

            NodeA.Layer.Connection_Add(this);

            RaiseDependant();
        }

        public PathfinderConnection(PathfinderConnection SourceConnection)
        {
            NodeA = SourceConnection.NodeA.ParentNode;
            NodeB = SourceConnection.NodeB.ParentNode;
            NodeA.Connection_Add(this, ref NodeA_ConnectionNum);
            NodeB.Connection_Add(this, ref NodeB_ConnectionNum);

            NodeA.Layer.Connection_Add(this);
            ValueCalc();
        }

        private void RemoveFromNodes()
        {
            NodeA.Connection_Remove(NodeA_ConnectionNum);
            NodeA = null;
            NodeA_ConnectionNum = -1;

            NodeB.Connection_Remove(NodeB_ConnectionNum);
            NodeB = null;
            NodeB_ConnectionNum = -1;
        }

        public PathfinderNode GetOtherNode(PathfinderNode Self)
        {
            if ( NodeA == Self )
            {
                return NodeB;
            }
            else
            {
                return NodeA;
            }
        }

        private void LinkIncrease()
        {
            LinkCount++;
        }

        private void LinkDecrease()
        {
            LinkCount--;
            if ( LinkCount == 0 )
            {
                Destroy();
            }
            else if ( LinkCount < 0 )
            {
                Debugger.Break();
            }
        }

        public void RaiseDependant()
        {
            PathfinderConnection tmpConnectionA = default(PathfinderConnection);

            if ( DependantConnection != null )
            {
                return;
            }

            if ( NodeA.ParentNode != NodeB.ParentNode )
            {
                if ( NodeA.ParentNode != null && NodeB.ParentNode != null )
                {
                    tmpConnectionA = NodeA.ParentNode.FindConnection(NodeB.ParentNode);
                    if ( tmpConnectionA == null )
                    {
                        DependantConnection = new PathfinderConnection(this);
                        DependantConnection.LinkIncrease();
                        DependantConnection.RaiseDependant();
                    }
                    else
                    {
                        DependantConnection = tmpConnectionA;
                        DependantConnection.LinkIncrease();
                    }
                }
            }
        }

        public void Destroy()
        {
            if ( Destroyed )
            {
                return;
            }
            Destroyed = true;

            PathfinderLayer Layer = NodeA.Layer;

            PathfinderNode tmpA = NodeA.ParentNode;
            PathfinderNode tmpB = NodeB.ParentNode;
            RemoveFromNodes();
            if ( tmpA != null )
            {
                tmpA.CheckIntegrity();
            }
            if ( tmpB != null && tmpB != tmpA )
            {
                tmpB.CheckIntegrity();
            }
            UnlinkParentDependants();
            Layer.Connection_Remove(Layer_ConnectionNum);
        }

        public void ForceDeallocate()
        {
            DependantConnection = null;
            NodeA = null;
            NodeB = null;
        }

        public void UnlinkParentDependants()
        {
            if ( DependantConnection != null )
            {
                PathfinderConnection tmpConnection = DependantConnection;
                DependantConnection = null;
                tmpConnection.LinkDecrease();
            }
        }

        public void ValueCalc()
        {
            if ( NodeA.Layer.Network_LayerNum == 0 )
            {
                Debugger.Break();
            }

            PathfinderNetwork.sFloodForValuesArgs Args = new PathfinderNetwork.sFloodForValuesArgs();
            int A = 0;
            int NumA = 0;
            PathfinderNetwork Map = NodeA.Layer.Network;
            float TotalValue = 0;
            Args.NodeValues = NodeA.Layer.Network.NetworkLargeArrays.Nodes_ValuesA;
            Args.FinishIsParent = false;
            Args.SourceNodes = NodeA.Layer.Network.NetworkLargeArrays.Nodes_Nodes;
            Args.SourceParentNodeA = NodeA;
            Args.SourceParentNodeB = NodeB;
            Args.CurrentPath = NodeA.Layer.Network.NetworkLargeArrays.Nodes_Path;
            Args.FinishNodeCount = NodeB.NodeCount;
            Args.FinishNodes = new PathfinderNode[NodeB.NodeCount];
            for ( A = 0; A <= NodeB.NodeCount - 1; A++ )
            {
                Args.FinishNodes[A] = NodeB.Nodes[A];
            }
            for ( NumA = 0; NumA <= NodeA.NodeCount - 1; NumA++ )
            {
                Args.CurrentPath.Nodes[0] = NodeA.Nodes[NumA];
                Args.CurrentPath.NodeCount = 1;
                for ( A = 0; A <= NodeA.NodeCount - 1; A++ )
                {
                    Args.NodeValues[NodeA.Nodes[A].Layer_NodeNum] = float.MaxValue;
                }
                for ( A = 0; A <= NodeB.NodeCount - 1; A++ )
                {
                    Args.NodeValues[NodeB.Nodes[A].Layer_NodeNum] = float.MaxValue;
                }
                Args.BestPaths = new PathfinderNetwork.Path[Args.FinishNodeCount];
                NodeA.Layer.Network.FloodForValues(Args);
                for ( A = 0; A <= NodeB.NodeCount - 1; A++ )
                {
                    if ( Args.BestPaths[A] == null )
                    {
                        Debugger.Break();
                        return;
                    }
                    TotalValue += Args.BestPaths[A].Value;
                }
            }
            Value = TotalValue / (NodeA.NodeCount * NodeB.NodeCount);
            if ( Value == 0.0F )
            {
                Debugger.Break();
            }

            CalcValueNum = -1;
        }
    }
}

File: reference/FlaMEVB/PathfinderConnection.vb
Public Class PathfinderConnection

    Public Destroyed As Boolean

    Public Layer_ConnectionNum As Integer = -1

    Public NodeA As PathfinderNode
    Public ReadOnly Property GetNodeA As PathfinderNode
        Get
            Return NodeA
        End Get
    End Property
    Public NodeA_ConnectionNum As Integer = -1
    Public ReadOnly Property GetNodeA_ConnectionNum As Integer
        Get
            Return NodeA_ConnectionNum
        End Get
    End Property

    Public NodeB As PathfinderNode
    Public ReadOnly Property GetNodeB As PathfinderNode
        Get
            Return NodeB
        End Get
    End Property
    Public NodeB_ConnectionNum As Integer = -1
    Public ReadOnly Property GetNodeB_ConnectionNum As Integer
        Get
            Return NodeB_ConnectionNum
        End Get
    End Property

    Private DependantConnection As PathfinderConnection 'the one above this that partially relies on this to exist

    Private LinkCount As Integer

    Public Value As Single = 1.0F
    Public ReadOnly Property GetValue As Single
        Get
            Return Value
        End Get
    End Property
    Public CalcValueNum As Integer = -1

    Public Sub New(NewNodeA As PathfinderNode, NewNodeB As PathfinderNode, NewValue As Single)

        If NewNodeA.Layer.Network_LayerNum > 0 Or NewNodeB.Layer.Network_LayerNum > 0 Or NewValue <= 0.0F Then
            Stop
            Exit Sub
        End If

        Value = NewValue

        LinkCount = 1

        NodeA = NewNodeA
        NodeB = NewNodeB
        NodeA.Connection_Add(Me, NodeA_ConnectionNum)
        NodeB.Connection_Add(Me, NodeB_ConnectionNum)

        NodeA.Layer.Connection_Add(Me)

        RaiseDependant()
    End Sub

    Public Sub New(SourceConnection As PathfinderConnection)

        NodeA = SourceConnection.NodeA.ParentNode
        NodeB = SourceConnection.NodeB.ParentNode
        NodeA.Connection_Add(Me, NodeA_ConnectionNum)
        NodeB.Connection_Add(Me, NodeB_ConnectionNum)

        NodeA.Layer.Connection_Add(Me)
        ValueCalc()
    End Sub

    Private Sub RemoveFromNodes()

        NodeA.Connection_Remove(NodeA_ConnectionNum)
        NodeA = Nothing
        NodeA_ConnectionNum = -1

        NodeB.Connection_Remove(NodeB_ConnectionNum)
        NodeB = Nothing
        NodeB_ConnectionNum = -1
    End Sub

    Public Function GetOtherNode(Self As PathfinderNode) As PathfinderNode

        If NodeA Is Self Then
            Return NodeB
        Else
            Return NodeA
        End If
    End Function

    Private Sub LinkIncrease()

        LinkCount += 1
    End Sub

    Private Sub LinkDecrease()

        LinkCount -= 1
        If LinkCount = 0 Then
            Destroy()
        ElseIf LinkCount < 0 Then
            Stop
        End If
    End Sub

    Public Sub RaiseDependant()
        Dim tmpConnectionA As PathfinderConnection

        If DependantConnection IsNot Nothing Then
            Exit Sub
        End If

        If NodeA.ParentNode IsNot NodeB.ParentNode Then
            If NodeA.ParentNode IsNot Nothing And NodeB.ParentNode IsNot Nothing Then
                tmpConnectionA = NodeA.ParentNode.FindConnection(NodeB.ParentNode)
                If tmpConnectionA Is Nothing Then
                    DependantConnection = New PathfinderConnection(Me)
                    DependantConnection.LinkIncrease()
                    DependantConnection.RaiseDependant()
                Else
                    DependantConnection = tmpConnectionA
                    DependantConnection.LinkIncrease()
                End If
            End If
        End If
    End Sub

    Public Sub Destroy()

        If Destroyed Then
            Exit Sub
        End If
        Destroyed = True

        Dim Layer As PathfinderLayer = NodeA.Layer

        Dim tmpA As PathfinderNode = NodeA.ParentNode
        Dim tmpB As PathfinderNode = NodeB.ParentNode
        RemoveFromNodes()
        If tmpA IsNot Nothing Then
            tmpA.CheckIntegrity()
        End If
        If tmpB IsNot Nothing And tmpB IsNot tmpA Then
            tmpB.CheckIntegrity()
        End If
        UnlinkParentDependants()
        Layer.Connection_Remove(Layer_ConnectionNum)
    End Sub

    Public Sub ForceDeallocate()

        DependantConnection = Nothing
        NodeA = Nothing
        NodeB = Nothing
    End Sub

    Public Sub UnlinkParentDependants()

        If DependantConnection IsNot Nothing Then
            Dim tmpConnection As PathfinderConnection = DependantConnection
            DependantConnection = Nothing
            tmpConnection.LinkDecrease()
        End If
    End Sub

    Public Sub ValueCalc()

        If NodeA.Layer.Network_LayerNum = 0 Then
            Stop
        End If

        Dim Args As New PathfinderNetwork.sFloodForValuesArgs
        Dim A As Integer
        Dim NumA As Integer
        Dim Map As PathfinderNetwork = NodeA.Layer.Network
        Dim TotalValue As Single
        Args.NodeValues = NodeA.Layer.Network.NetworkLargeArrays.Nodes_ValuesA
        Args.FinishIsParent = False
        Args.SourceNodes = NodeA.Layer.Network.NetworkLargeArrays.Nodes_Nodes
        Args.SourceParentNodeA = NodeA
        Args.SourceParentNodeB = NodeB
        Args.CurrentPath = NodeA.Layer.Network.NetworkLargeArrays.Nodes_Path
        Args.FinishNodeCount = NodeB.NodeCount
        ReDim Args.FinishNodes(NodeB.NodeCount - 1)
        For A = 0 To NodeB.NodeCount - 1
            Args.FinishNodes(A) = NodeB.Nodes(A)
        Next
        For NumA = 0 To NodeA.NodeCount - 1
            Args.CurrentPath.Nodes(0) = NodeA.Nodes(NumA)
            Args.CurrentPath.NodeCount = 1
            For A = 0 To NodeA.NodeCount - 1
                Args.NodeValues(NodeA.Nodes(A).Layer_NodeNum) = Single.MaxValue
            Next
            For A = 0 To NodeB.NodeCount - 1
                Args.NodeValues(NodeB.Nodes(A).Layer_NodeNum) = Single.MaxValue
            Next
            ReDim Args.BestPaths(Args.FinishNodeCount - 1)
            NodeA.Layer.Network.FloodForValues(Args)
            For A = 0 To NodeB.NodeCount - 1
                If Args.BestPaths(A) Is Nothing Then
                    Stop
                    Exit Sub
                End If
                TotalValue += Args.BestPaths(A).Value
            Next
        Next
        Value = TotalValue / (NodeA.NodeCount * NodeB.NodeCount)
        If Value = 0.0F Then
            Stop
        End If

        CalcValueNum = -1
    End Sub
End Class